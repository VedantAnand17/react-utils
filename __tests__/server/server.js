import _ from 'lodash';
import supertest from 'supertest';

import serverFactory from 'server/server';

jest.mock('serve-favicon', () => jest.fn(() => (req, res, next) => next()));

jest.mock('webpack', () => {
  const mock = () => ({
    apply: jest.fn(),
    plugin: jest.fn(),
    watch: jest.fn(),
  });
  mock.ProgressPlugin = Object;
  return mock;
});

jest.mock('webpack-hot-middleware',
  () => jest.fn(() => (req, res, next) => next()));

const TEST_CONTEXT = `${__dirname}/test_data`;

const TEST_FAVICON_PATH = '/path/to/favicon.ico';

const TEST_WEBPACK_CONFIG = {
  context: TEST_CONTEXT,
  output: {
    path: '/path/to/assets',
    publicPath: '/test/public/path/',
  },
};

afterEach(() => {
  delete process.env.DEV_TOOLS;
  jest.resetAllMocks();
});

test('Favicon support', () => {
  const server = serverFactory(TEST_WEBPACK_CONFIG, {
    favicon: _.clone(TEST_FAVICON_PATH),
    logger: console,
  });
  expect(require('serve-favicon')).toHaveBeenCalledWith(TEST_FAVICON_PATH);
  return server;
});

test('Launch with dev tools', () => {
  process.env.DEV_TOOLS = true;
  const server = serverFactory(TEST_WEBPACK_CONFIG, { logger: console });
  return server;
});

describe('Server is functional', () => {
  let server;

  beforeAll(async () => {
    server = supertest(
      // TODO: I guess, it is better to completely mock the logger here,
      // to not print anything into console during the normal test pass.
      await serverFactory(TEST_WEBPACK_CONFIG, {
        cspSettingsHook: (csp) => {
          csp.directives['default-src'].push('https://sample.url');
          return csp;
        },
        logger: console,
      }),
    );
  });

  test('Simple request', async () => {
    await server.get('/').expect(200)
      .expect((res) => {
        expect(res.headers['content-security-policy']).toMatchSnapshot();
      });
    await server.get('/').expect(200)
      .expect((res) => {
        expect(res.headers['content-security-policy']).toMatchSnapshot();
      });
  });
});
