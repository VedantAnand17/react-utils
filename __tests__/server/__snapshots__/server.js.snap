// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Server is functional Simple request 1`] = `"default-src 'self' https://sample.url;base-uri 'self';block-all-mixed-content;font-src 'self' data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self' 'unsafe-eval' 'nonce-00000000-0000-0000-0000-000000000001';script-src-attr 'none';style-src 'self' 'unsafe-inline';frame-src 'self' https://*.youtube.com;upgrade-insecure-requests"`;

exports[`Server is functional Simple request 2`] = `"default-src 'self' https://sample.url;base-uri 'self';block-all-mixed-content;font-src 'self' data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self' 'unsafe-eval' 'nonce-00000000-0000-0000-0000-000000000002';script-src-attr 'none';style-src 'self' 'unsafe-inline';frame-src 'self' https://*.youtube.com;upgrade-insecure-requests"`;
